src/btc_handler.mo

```diff
@@ -0,0 +1,204 @@
+      spender : { owner : Principal; subaccount : ?Blob };
+      amount : Nat;
+      expected_allowance : ?Nat;
+      expires_at : ?Nat64;
+      fee : ?Nat;
+      memo : ?Blob;
+      created_at_time : ?Nat64;
+    }) -> async { #Ok : Nat; #Err : ApproveError };
+  };
+
+  type CkbtcMinter = actor {
+    get_btc_address : shared ({ owner : ?Principal; subaccount : ?Blob }) -> async Text;
+    update_balance : shared ({ owner : ?Principal; subaccount : ?Blob }) -> async {
+      #Ok : [UtxoStatus];
+      #Err : UpdateBalanceError;
+    };
+    estimate_withdrawal_fee : shared query ({ amount : ?Nat64 }) -> async ({
+      bitcoin_fee : Nat64;
+      minter_fee : Nat64;
+    });
+    get_deposit_fee : shared query () -> async Nat64;
+    retrieve_btc_with_approval : shared ({
+      address : Text;
+      amount : Nat64;
+      from_subaccount : ?Blob;
+    }) -> async {
+      #Ok : { block_index : Nat64 };
+      #Err : RetrieveBtcWithApprovalError;
+    };
+    retrieve_btc_status_v2 : shared query ({ block_index : Nat64 }) -> async RetrieveBtcStatusV2;
+  };
+
+  public type NotifyError = UpdateBalanceError or { #NotMinted };
+
+  public class BtcHandler(auctionPrincipal : Principal, ckbtcLedgerPrincipal : Principal, ckbtcMinterPrincipal : Principal) {
+
+    let ckbtcLedger : CkbtcLedger = actor (Principal.toText(ckbtcLedgerPrincipal));
+    let ckbtcMinter : CkbtcMinter = actor (Principal.toText(ckbtcMinterPrincipal));
+
+    public func getDepositAddress(p : Principal) : async* Text {
+      await ckbtcMinter.get_btc_address({
+        owner = ?auctionPrincipal;
+        subaccount = ?TokenHandler.toSubaccount(p);
+      });
+    };
+
+    public func notify(p : Principal) : async* R.Result<(), NotifyError> {
+      let resp = await ckbtcMinter.update_balance({
+        owner = ?auctionPrincipal;
+        subaccount = ?TokenHandler.toSubaccount(p);
+      });
+      switch (resp) {
+        case (#Err err) return #err(err);
+        case (#Ok utxos) {
+          for (utxo in utxos.vals()) {
+            switch (utxo) {
+              case (#Minted _) {};
+              case (_) return #err(#NotMinted);
+            };
+          };
+        };
+      };
+      #ok();
+    };
+
+    public func withdraw(address : Text, amount : Nat, expected_fee : ?Nat) : async* {
+      #Ok : { block_index : Nat64 };
+      #Err : ApproveError or RetrieveBtcWithApprovalError;
+    } {
+      let approveRes = await ckbtcLedger.icrc2_approve({
+        from_subaccount = null;
+        amount;
+        spender = { owner = ckbtcMinterPrincipal; subaccount = null };
+        fee = expected_fee;
+        expected_allowance = null;
+        created_at_time = null;
+        expires_at = null;
+        memo = null;
+      });
+      switch (approveRes) {
+        case (#Err err) #Err(err);
+        case (#Ok allowanceAmount) {
+          await ckbtcMinter.retrieve_btc_with_approval({
+            address;
+            amount = Nat64.fromNat(allowanceAmount);
```

Comment:
I suppose here we have to reduce the amount by fee.

Reply:
Right